framework/include/materials/Material.h

```diff
@@ -258,6 +259,25 @@ class Material :
    */
   virtual QpData * createData();
 
+  /**
+   * Get a apecial MaterialUserObject which can be explicitly called by the material
+   * during the material property calculation.
+   * @param name The name of the parameter key of the user object to retrieve
+   * @return The user object with name associated with the parameter 'name'
+   */
+  template<class T>
+  const T & getMaterialUserObject(const std::string & name);
```

Comment:
Andrew, I wrote this in a different issue, but as dynamic_cast does not work with a mere forward declare your suggestion is not trivial to implement! I may go down that route anyhow if I figure out a way....